### PR TITLE
Pass static argument to scene2d

### PR DESCRIPTION
**Pull request by dfcreative**
Merged along with this https://github.com/gl-vis/gl-plot2d/pull/4 will remove image of picked data from rendered result

### CODE REVIEW

**Comment by etpinard** [comment]
Sweet! Let me test it out.

**Comment by etpinard** [comment]
@dfcreative I've got some news! Looks like you're almost there.

Most gl2d [mocks](https://github.com/plotly/plotly.js/tree/master/test/image/mocks) can now be generated in a reproducible way!

Here's the list of mocks that are still problematic:

image

By looking at them quickly, all mocks in :arrow_double_up: have made up of one or more `gl-line2d` object. Maybe we'll need to bump the `gl-shader` dep there too?

**Comment by etpinard** [comment]
For reference:

image

**Comment by etpinard** [comment]
Wait @dfcreative does gl-plot2d need `gl-shader@4.2.1` too?

**Comment by etpinard** [comment]
Ok I just checked with

image


and more or less the same mocks are failing intermittently:

image

**Comment by dfcreative** [comment]
Yeah, eventually gl-shader@4.2.1 fixes errors in console, but resulting images still have bad dimensions, due to wrong canvas size etc, gotta check that the other day.

**Comment by dfcreative** [comment]
Ok, 2558160ce427665a5c389d6e2360371126a50105 fixes dimensions, so that initial case works. Sometimes though, `toImage` in some reason is invoked before `plot`, which skews everything.

**Comment by etpinard** [comment]
@dfcreative great. I'll test this out this afternoon! Thanks!!

File: src/plots/gl2d/scene2d.js [hunk 1]
```diff
@@ -169,6 +169,26 @@ proto.toImage = function(format) {
...
+        fullLayout = this.fullLayout;
+
+    var width = fullLayout.width,
+        height = fullLayout.height,
+        pixelWidth = Math.ceil(pixelRatio * width) |0,
+        pixelHeight = Math.ceil(pixelRatio * height) |0;
+
+    // check for resize
+    var canvas = this.canvas;
+    if(canvas.width !== pixelWidth || canvas.height !== pixelHeight) {
```
Comment (dfcreative) [inline src/plots/gl2d/scene2d.js]: This is duplicate of https://github.com/dfcreative/plotly.js/blob/07af2b83e8d6b4bb9da5363ac27ffeeebadf0fd0/src/plots/gl2d/scene2d.js#L131 and https://github.com/dfcreative/plotly.js/blob/07af2b83e8d6b4bb9da5363ac27ffeeebadf0fd0/src/plots/gl2d/scene2d.js#L374. Should I put it into a method `updateSize`?

Comment (dfcreative) [inline src/plots/gl2d/scene2d.js]: Ok, refactored so that now there is `updateSize` method

**Comment by etpinard** [comment]
@dfcreative looking good!

See https://github.com/plotly/plotly.js/commit/052dc636003b9393199b26d484478a3476adf4e0 where I re-generated every gl2d test image. I think they all look good now.

I'm still getting some run-to-run inconsistencies when generating images in batch (we use a concurrency of 5 in our image test routine). **BUT** generating them one-after-another works like charm :tada: (at least of my machine). In commit https://github.com/plotly/plotly.js/commit/b321cb9a5f79b60c48fa5e929277bb5db764663a we'll see if that same is true on CircleCI.

**Comment by dfcreative** [comment]
What sort of inconsistencies?

The inconsistency I faced with is caused by occasional delay of initial `scene.plot` method, which in some reason sometimes is called after `scene.toImage`, so that rendered images have uninitialized content, same as real plots at the moment.
I did not manage to consistently reproduce that use-case, my laptop acclerates and initial delay becomes insufficient)

**Comment by etpinard** [comment]
> What sort of inconsistencies?

Sorry, I should've been clearer. For example, `gl2d_10` when ran in batch yield

image

the baseline is on the left, the generated image on the right.

**Comment by etpinard** [comment]
On CiricleCI, these mocks

image

are failing the image comparison test when ran in queue (i.e. with zero concurrency).

Note that the CircleCI machine is a little weaker than most laptops.

**Comment by etpinard** [comment]
@dfcreative having to run the gl2d image tests with zero concurrency would be a great start. The only drawback would be slower test runs - which isn't that big of a deal at the moment.

So, figuring out why some gl2d mocks are having issues on CircleCI would be amazing.

**Comment by dfcreative** [comment]
Yes, I will check it out. Also the inconsistencies - these are the same I mentioned.

**Comment by etpinard** [comment]
> Also the inconsistencies - these are the same I mentioned.

Great. Thanks for the info. :+1:

**Comment by dfcreative** [comment]
Hm. These examples, except for the _gl2d_pointcloud-basic_, look exactly the same as real plots. May I ask how do you get baseline images? They seem a bit taller than the rendered ones.
Baseline: 
gl2d_simple_inset
Rendered:
image

**Comment by etpinard** [comment]
> May I ask how do you get baseline images?

Full docs :arrow_right:  https://github.com/plotly/plotly.js/tree/master/test/image#c-generate-or-update-existing-baseline-image

**Comment by dfcreative** [comment]
The problem is that the baseline images are taller than the default plot height I get (500px vs 450px). These mocks do not have sizes in `layout`, so I guess streambed may just preset own defaults or something like that.
If to preset dimensions in mocks:
```json
"layout": {
...
        "width": 700,
        "height": 500
}
```
Everything works fine. Shall we?

**Comment by etpinard** [comment]
@dfcreative 

Hmm, looks like CircleCI is still having issues - see https://circleci.com/gh/plotly/plotly.js/2597

**Comment by dfcreative** [comment]
@monfera hi!

There seems to be a single mock failing, that is [gl2d_pointcloud-basic](https://github.com/plotly/plotly.js/blob/master/test/image/mocks/gl2d_pointcloud-basic.json#L10).

The problem is when we try to read pixels from webgl in static canvas with disabled blending, they get mixed with gray-ish background or something like that.
Expected:
image
Actual:
image

What are the use-cases we may want to disable blend [here](https://github.com/plotly/plotly.js/blob/master/test/image/mocks/gl2d_pointcloud-basic.json#L10)? If we need solid color we can set `"blend": true` by default and set `"color": "rgba(255, 0, 0, 1)"` for example.

**Comment by etpinard** [comment]
On CircleCI, `gl2d_multiple_subplots` is generated as

image

**Comment by monfera** [comment]
@dfcreative sorry to hear it's causing testing trouble. The blending is disabled [here](https://github.com/gl-vis/gl-pointcloud2d/blob/master/pointcloud2d.js#L165-L171], the purpose was an approx. 2x speedup on shader performance, which mattered a lot on the 1m point user requirement case. Admittedly the comment could be more specific. It should be possible to specify `blend: true` in the [attributes](https://github.com/plotly/plotly.js/blob/master/src/traces/pointcloud/attributes.js#L76-L86) to let this test pass. Though I had image test running problems last time, I can help with a small PR enabling blending, please let me know if I should go ahead with it.

**Comment by dfcreative** [comment]
@monfera I would guess practically here can be two cases: if there is little data, then having blending enabled will give better results, whereas if there is 1m points, the precision of a single point does not matter and blending is better to be off.

**Comment by dfcreative** [comment]
@monfera is it ok to detect blending like that https://github.com/plotly/plotly.js/pull/980/commits/0efb54ed4886ae3fe322392602d738d257e8f41f?
I am not sure about magic numbers, the guess is if there are more than 100 points then disable blend.

**Comment by monfera** [comment]
@dfcreative it's possible as we already have a handle on [how many points](https://github.com/gl-vis/gl-pointcloud2d/blob/master/pointcloud2d.js#L133) are visible at any one time. (PS. just saw your commit which looks for the total number of points, and in `gl-scatter2d` we have the point count for the currently visible, on-screen points.)

One thing though, there _can_ be a big difference in the rendered output even if there are lots of points. With proper blending, it's possible to have a [density cloud effect](http://stackoverflow.com/questions/14271584/r-legend-for-color-density-scatterplot-produced-using-smoothscatter) but of course it comes with shader time expense.

File: src/traces/pointcloud/attributes.js [hunk 2]
```diff
@@ -75,7 +75,7 @@ module.exports = {
         },
         blend: {
             valType: 'boolean',
-            dflt: false,
+            dflt: undefined,
```
Comment (etpinard) [inline src/traces/pointcloud/attributes.js]: we usually use `null` in situation like this one

File: src/traces/pointcloud/convert.js [hunk 3]
```diff
@@ -174,7 +174,12 @@ proto.updateFast = function(options) {
 
     markerColor[3] *= opacity;
     this.pointcloudOptions.color = markerColor;
-    this.pointcloudOptions.blend = options.marker.blend;
+
+    var blend = options.marker.blend;
+    if (blend == null) {
```
Comment (etpinard) [inline src/traces/pointcloud/convert.js]: ... and `(blend === null)`

Comment (dfcreative) [inline src/traces/pointcloud/convert.js]: Done

**Comment by etpinard** [comment]
Merging.

I'll add the CI test command in a separate PR.

**Pull request merged**